cmd/access/node_builder/upstream_connector.go

```diff
@@ -0,0 +1,124 @@
+package node_builder
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/lifecycle"
+	"github.com/onflow/flow-go/network/p2p"
+)
+
+// upstreamConnector tries to connect the unstaked AN with atleast one of the configured bootstrap access nodes
+type upstreamConnector struct {
+	lm                  *lifecycle.LifecycleManager
+	bootstrapIdentities flow.IdentityList
+	logger              zerolog.Logger
+	unstakedNode        *p2p.Node
+	ctx                 context.Context
+	cancel              context.CancelFunc
+}
+
+func newUpstreamConnector(ctx context.Context, bootstrapIdentities flow.IdentityList, unstakedNode *p2p.Node, logger zerolog.Logger) *upstreamConnector {
+
+	ctx, cancel := context.WithCancel(ctx)
+	return &upstreamConnector{
+		lm:                  lifecycle.NewLifecycleManager(),
+		bootstrapIdentities: bootstrapIdentities,
+		unstakedNode:        unstakedNode,
+		logger:              logger,
+		ctx:                 ctx,
+		cancel:              cancel,
+	}
+}
+func (connector *upstreamConnector) Ready() <-chan struct{} {
+	connector.lm.OnStart(func() {
+
+		bootstrapPeerCnt := len(connector.bootstrapIdentities)
+		resultChan := make(chan result, bootstrapPeerCnt)
+		defer close(resultChan)
+
+		// a shorter context for the connection worker
+		ctx, cancel := context.WithTimeout(connector.ctx, 5*time.Second)
```

Comment:
Can you address https://github.com/onflow/flow-go/pull/1098#discussion_r685363015?

Reply:
left a comment there.

Reply:
@vishalchangrani this is what I meant: https://github.com/onflow/flow-go/pull/1127

Reply:
merged it in thanks